Enable API resource customization using distinct objPath and dumpPath

tmishina opened the pull request:
## Description

This PR  is the successor of #7928 and https://github.com/openshift/compliance-operator/pull/788.

The final contents of #7928 leveraged [the Compliance Operator's template substitution feature](https://github.com/openshift/compliance-operator/pull/750), but its format force rule authors write both default path (in `.template.vars.filepath`) and customized path placeholder (in `warnings`), which may increase rule authoring workload and update inconsistency.

This PR solves the issue by modifying the OpenShift cluster setting macros.  Now the macros can accept both default and customized path, and it provides simpler authoring format for the rule authors.

(Please also see the description of #7928)

#### Rationale:

- This PR expands the capability of ComplianceAsCode/content and Compliance Operator to various kinds of installations including managed OpenShift Services

tmishina commented:
Currently, the branch contains just a revised macro and example implementation for one rule. I confirmed that this modification can generate expected XML rules, but it's still a work in progress.

Vincent056 reviewed (approved):
Thanks for adding the PR, I think this looks good to me.

Vincent056 commented:
So for  `openshift_filtered_path()`, I think we just need to add one more parameter (dump location) to it.

tmishina commented on applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:
```diff
@@ -40,15 +45,15 @@ ocil: |-
     ocp_data: "true"
     filepath: |-
-      {{{ openshift_filtered_path('/api/v1/namespaces/openshift-kube-apiserver/configmaps/config', jqfilter) }}}
+      {{{ openshift_filtered_path(apipathdefault, jqfilter) }}}
```
@Vincent056  Could you tell me why you think `openshift_filtered_path(path, filter)` requires additional parameter? In my understanding, the return value of `openshift_filtered_path()` will be used by `oscap` command, and therefore only default (dump) path is required as I wrote here.

Vincent056 replied:
Hi @tmishina Sorry for the confusion, it was meant to say `openshift_cluster_setting`

Mab879 commented:
/ok-to-test

Vincent056 commented:
/retest

Vincent056 commented on applications/openshift/api-server/api_server_admission_control_plugin_AlwaysAdmit/rule.yml:
```diff
@@ -4,6 +4,12 @@ prodtype: ocp4
 
+{{% set custom_jqfilter = '[.data."{{.var_openshift_kube_apiserver_config_data_name}}" | fromjson]' %}}
+{{% set default_jqfilter = '[.data."config.yaml" | fromjson]' %}}
+{{% set custom_api_path = '/api/v1/namespaces/{{.var_openshift_kube_apiserver_namespace}}/configmaps/{{.var_openshift_kube_apiserver_config}}' %}}
```
I am wondering if we want to set `var_openshift_kube_apiserver_namespace` as something like `var_cluster_namespace` instead since it will be reused by many other rules? 

```
custom_api_path = '/api/v1/namespaces{{.var_cluster_namespace}}configmaps/{{.var_openshift_kube_apiserver_config}}'
```

And Set default value for `var_cluster_namespace` as `/`.

Otherwise I think it looks good/

Vincent056 replied:
I guess as for now, we need to have a tailored profile to keep track of the rules that have completely different endpoints like this one

Vincent056 commented on shared/macros.jinja:
```diff
@@ -38,27 +38,56 @@ Therefore, you need to use a tool that can query the OCP API, retrieve the {{% i
-    file.
-  </li>
-{{% endfor %}}
+    {{% for obj_path, v in arg.items() %}}
```
I didn't think of this one, this is a nice way to be backward comptitabable

tmishina commented:
### Bulk change for api-server rules

|               | default                                                      | custom                                                       |
| ------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
| resource path | `/api/v1/namespaces/openshift-kube-apiserver/configmaps/config` | `/api/v1/namespaces/{{.var_openshift_kube_apiserver_namespace}}/configmaps/{{.var_openshift_kube_apiserver_config}}` |
| jqfilter      | `.data."config.yaml" \|fromjson`                              | `.data."{{.var_openshift_kube_apiserver_config_data_name}}" \|fromjson` |

### Bulk change for kube controller manager rules

|               | default                                                      | custom                                              |
| ------------- | ------------------------------------------------------------ | --------------------------------------------------- |
| resource path | `/api/v1/namespaces/openshift-kube-controller-manager/configmaps/config` | `{{.var_kube_controller_manager_config_filepath}}`  |
| jqfilter      | `[.data."config.yaml" \|fromjson \|if .extendedArguments["port"]!=null then .extendedArguments["port"]==["0"] else true end]` (example) | `{{.var_kube_controller_manager_port_zero_filter}}` |

For controller manager rules, different jqfilter is configured for each rule.

### Bulk change for etcd rules

|               | default                                                 | custom                          |
| ------------- | ------------------------------------------------------- | ------------------------------- |
| resource path | `/api/v1/namespaces/openshift-etcd/configmaps/etcd-pod` | `{{.var_etcd_filepath}`         |
| jqfilter      | `[.data."pod.yaml"]`                                    | `{{.var_etcd_argument_filter}}` |

The reason why custom value is `{{.var}}` instead of `/api/v1/namespaces/{{.var}}` is that the path is completely different from default. Actual value for HyperShift is `/api/v1/namespaces/master-tmishina1/pods?labelSelector=app%3Detcd`. jqfilter for HyperShift is `[.items[0].spec.containers[0].command | join(" ")]`

### Bulk change for kube scheduler rules

|               | default                                                      | custom                               |
| ------------- | ------------------------------------------------------------ | ------------------------------------ |
| resource path | `/api/v1/namespaces/openshift-kube-scheduler/configmaps/kube-scheduler-pod` | `{{.var_scheduler_filepath}}`        |
| jqfilter      | `[.data."pod.yaml"]`                                         | `{{.var_scheduler_argument_filter}}` |

As similar to etcd rules, custom values are completely different.  HyperShift resource path is `/api/v1/namespaces/master-tmishina1/pods?labelSelector=app%3Dkube-scheduler`, and jqfilter is `[.items[0].spec.containers[0].args | join(" ")]`

### Rule change with value change

- api_server_bind_address
  - original: `bindAddress` must be `0.0.0.0:6443`
  - updated: `bindAddress` must match pattern `.+`
  - reason: `bindAddress` must be set, but the port number (`6443`) doesn't matter.
- Cert and key rules (api_server_client_ca, api_server_etcd_ca, api_server_kubelet_certificate_authority, api_server_kubelet_client_cert, api_server_service_account_public_key, api_server_tls_cert, api_server_tls_private_key, controller_service_account_ca, controller_service_account_private_key, etcd_cert_file, etcd_key_file, etcd_peer_cert_file, etcd_peer_key_file, kubelet_configure_tls_cert, kubelet_configure_tls_key)
  - original: cert or key file must be equals to constant like `/etc/kubernetes/static-pod-certs/configmaps/client-ca/ca-bundle.crt`
  - updated: cert or key file must match pattern `.+`
  - reason: cert or key file must be set, but its value doesn't matter. Any cert file can be accepted.

tmishina commented:
@Vincent056 I completed rule changes. 54 rules are changed, and I summarized the changes above. 

Now I have several discussion points and questions; could you take a time to think about the following items? Thank you very much for your help.

1. is `.var_cluster_namespace` better than `.var_openshift_kube_apiserver_namespace`? 
  - [You suggested using var_cluster_namespace](https://github.com/ComplianceAsCode/content/pull/8231#discussion_r818267690)  and I thought about it. As wrote in the summary, master namespace name is used only in the kube API server rules. In other rules, master namespace name is included in other variables (example: `var_etcd_filepath` is `/api/v1/namespaces/master-tmishina1/pods?labelSelector=app%3Detcd`, containing the master name `master-tmishina1`). In this situation, I think it is better to simply keep using `.var_openshift_kube_apiserver_namespace`.  Could you give me your thoughts again?
2. How to create a profile with custom values?
  - I agree [your comment](https://github.com/ComplianceAsCode/content/pull/8231#discussion_r818268282) that it is good to provide a tailored profile. I found several tailored profiles in `products/ocp4/profiles`, but I can't find a profile which contains `setValue`-like description. Could you point to document for the profile?
3. Are the changes described in "Rule change with value change" valid?
- I verified that think these changes does not breach backward compatibility, but are these changes valid with respect to the CIS Benchmark?

tmishina commented:
@Vincent056 now several tests fail. Does my code cause these errors?  Rules that I didn't changed also failed, and therefore I imagine some systematic error happens.  I'd appreciate it if you can help solving this situation!

Vincent056 commented:
> @Vincent056 now several tests fail. Does my code cause these errors? Rules that I didn't changed also failed, and therefore I imagine some systematic error happens. I'd appreciate it if you can help solving this situation!

Let me take a look

Vincent056 commented:
> @Vincent056 now several tests fail. Does my code cause these errors? Rules that I didn't changed also failed, and therefore I imagine some systematic error happens. I'd appreciate it if you can help solving this situation!

@tmishina have you tried to rebase it, the CI on other PRs looks fine

tmishina commented:
/retest

tmishina commented:
/retest

tmishina commented:
@Vincent056 I did `git rebase master` using the up-to-date master branch of the origin (https://github.com/ComplianceAsCode/content), but the tests still fail.

What is the version of Compliance Operator that this CI pipeline uses? This PR requires functionaliries of https://github.com/openshift/compliance-operator/pull/800 and https://github.com/openshift/compliance-operator/pull/750, which have already been merged but have not been included in the latest release of Compliance Operator.

Vincent056 commented on applications/openshift/api-server/api_server_admission_control_plugin_AlwaysAdmit/rule.yml:
```diff
@@ -4,6 +4,12 @@ prodtype: ocp4
 
 title: 'Disable the AlwaysAdmit Admission Control Plugin'
 
+{{% set custom_jqfilter = '[.data."{{.var_openshift_kube_apiserver_config_data_name}}" | fromjson]' %}}
```
It seems like from the log, `{{.var_openshift_kube_apiserver_config_data_name}}` did not get parsed

Vincent056 replied:
We might want to wait for a CO upstream release

Vincent056 commented:
> 

It uses CO upstream image, we need to do an upstream release to include the change.

Vincent056 commented:
> @Vincent056 I did `git rebase master` using the up-to-date master branch of the origin (https://github.com/ComplianceAsCode/content), but the tests still fail.
> 
> What is the version of Compliance Operator that this CI pipeline uses? This PR requires functionaliries of [openshift/compliance-operator#800](https://github.com/openshift/compliance-operator/pull/800) and [openshift/compliance-operator#750](https://github.com/openshift/compliance-operator/pull/750), which have already been merged but have not been included in the latest release of Compliance Operator.

It uses CO upstream image, we need to do an upstream release to include the change.

tmishina commented:
/retest

tmishina commented:
/retest

tmishina commented:
@Vincent056 I modified [the catalog source](https://github.com/ComplianceAsCode/content/pull/8231/files#diff-860809d134b44534ae8b16ad2ffa2e3beb88556f32c9124358744ad37b614d2a) to temporarily use `master` branch of Compliance Operator [image](https://quay.io/repository/osco4hypershift/compliance-operator) and [bundle](https://quay.io/repository/osco4hypershift/compliance-operator-bundle) instead of latest release.  As a result, the tests are passed.

Vincent056 commented:
> @Vincent056 I modified [the catalog source](https://github.com/ComplianceAsCode/content/pull/8231/files#diff-860809d134b44534ae8b16ad2ffa2e3beb88556f32c9124358744ad37b614d2a) to temporarily use `master` branch of Compliance Operator [image](https://quay.io/repository/osco4hypershift/compliance-operator) and [bundle](https://quay.io/repository/osco4hypershift/compliance-operator-bundle) instead of latest release. As a result, the tests are passed.

That's really nice! thanks for testing it out, I think we can merge it after a new CO release next week

Vincent056 commented:
> @Vincent056 I completed rule changes. 54 rules are changed, and I summarized the changes above.
> 
> Now I have several discussion points and questions; could you take a time to think about the following items? Thank you very much for your help.
> 
>     1. is `.var_cluster_namespace` better than `.var_openshift_kube_apiserver_namespace`?
> 
> 
>     * [You suggested using var_cluster_namespace](https://github.com/ComplianceAsCode/content/pull/8231#discussion_r818267690)  and I thought about it. As wrote in the summary, master namespace name is used only in the kube API server rules. In other rules, master namespace name is included in other variables (example: `var_etcd_filepath` is `/api/v1/namespaces/master-tmishina1/pods?labelSelector=app%3Detcd`, containing the master name `master-tmishina1`). In this situation, I think it is better to simply keep using `.var_openshift_kube_apiserver_namespace`.  Could you give me your thoughts again?
> 
> 
>     2. How to create a profile with custom values?
> 
> 
>     * I agree [your comment](https://github.com/ComplianceAsCode/content/pull/8231#discussion_r818268282) that it is good to provide a tailored profile. I found several tailored profiles in `products/ocp4/profiles`, but I can't find a profile which contains `setValue`-like description. Could you point to document for the profile?
> 
> 
>     3. Are the changes described in "Rule change with value change" valid?
> 
> 
>     * I verified that think these changes does not breach backward compatibility, but are these changes valid with respect to the CIS Benchmark?

1. Thanks for the explanation, In that case, I think your original naming looks better.

2. Here is an example of tailored profile setting a value, for more you can find it here: https://docs.openshift.com/container-platform/4.6/security/compliance_operator/compliance-operator-tailor.html
```yaml
apiVersion: compliance.openshift.io/v1alpha1
kind: TailoredProfile
metadata:
  name: test-tailor-profile
spec:
  extends: rhcos4-moderate
  title: test-tailor-profile
  setValues:
  - name: rhcos4-sshd-idle-timeout-value
    rationale: test for a custom value
    value: '3600'
  description: test-profile
```

3. Since all the CI passed, the change should have backward compatibility. As for `Rule change with value change` I think they all makes sense to me.

tmishina commented:
@Vincent056 Thank you for your review. Regarding comment 2, I shared an example tailored profile.

https://gist.github.com/tmishina/56d71dedc2409a5d797628a43483143d

Is there any good location rather than my personal gist to share this profile?

Vincent056 commented:
> @Vincent056 Thank you for your review. Regarding comment 2, I shared an example tailored profile.
> 
> https://gist.github.com/tmishina/56d71dedc2409a5d797628a43483143d
> 
> Is there any good location rather than my personal gist to share this profile?

I think maybe we could add a doc to the CO repo on how to use CO on HyperShift?

tmishina commented:
/retest

tmishina commented:
/retest

tmishina commented:
/retest

tmishina commented:
/retest

Vincent056 commented:
/retest all

Vincent056 commented:
/test all

tmishina commented:
/retest

Vincent056 commented:
/retest

Vincent056 commented:
/retest

tmishina commented:
/retest

Vincent056 commented:
/retest

tmishina commented:
/retest

Vincent056 reviewed (approved):
all tests passed, the error is not relevant to this PR
